NetKVM/Common/ParaNdis_TX.cpp

```diff
@@ -1071,3 +1139,145 @@ ULONG CNB::Copy(PVOID Dst, ULONG Length) const
 
     return Copied;
 }
+
+ULONG CNB::Copy(PVOID Dst, ULONG Length, ULONG DataOffset) const
+{
+    //find the MDL where the offset position, DataOffset, is located.
+    ULONG Offset = 0;
+    PMDL DataOffsetMDL = nullptr;
+
+    for (DataOffsetMDL = NET_BUFFER_CURRENT_MDL(m_NB);; DataOffsetMDL = DataOffsetMDL->Next)
+    {
+        Offset += MmGetMdlByteCount(DataOffsetMDL);
+        if (Offset > DataOffset + NET_BUFFER_CURRENT_MDL_OFFSET(m_NB) || DataOffsetMDL->Next == nullptr)
+        {
+            break;
+        }
+    }
+
+    ULONG CurrOffset = MmGetMdlByteCount(DataOffsetMDL) - (Offset - DataOffset - NET_BUFFER_CURRENT_MDL_OFFSET(m_NB));
+    ULONG Copied = 0;
+
+    Length = min(Length, NET_BUFFER_DATA_LENGTH(m_NB) - DataOffset);
+
+    for (PMDL CurrMDL = DataOffsetMDL;
+        CurrMDL != nullptr && Copied < Length;
+        CurrMDL = CurrMDL->Next)
+    {
+        ULONG CurrLen;
+        PVOID CurrAddr = nullptr;
+
+        NdisQueryMdl(CurrMDL, &CurrAddr, &CurrLen, MM_PAGE_PRIORITY(LowPagePriority | MdlMappingNoExecute));
+
+        if (CurrAddr == nullptr)
+        {
+            break;
+        }
+
+        CurrLen = min(CurrLen - CurrOffset, Length - Copied);
+
+        NdisMoveMemory(RtlOffsetToPointer(Dst, Copied),
+            RtlOffsetToPointer(CurrAddr, CurrOffset),
+            CurrLen);
+        Copied += CurrLen;
+        CurrOffset = 0;
+    }
+
+    return Copied;
+}
+
+#ifdef MAX_FRAGMENTS_IN_ONE_NB
+    bool CNB::AllocateAndFillCopySGL() 
+    {
+        //decide how many pages should be used 
+        ULONG uDataLength = GetDataLength();
+        ULONG uPages = (GetDataLength() + PAGE_SIZE - 1) / PAGE_SIZE;
+
+        m_ExtraNBStorage = (CExtendedNBStorage*)NdisAllocateMemoryWithTagPriority(m_Context->MiniportHandle,
+            sizeof(CExtendedNBStorage),
+            PARANDIS_MEMORY_TAG,
+            NormalPoolPriority);
+
+        if (m_ExtraNBStorage == nullptr)
+        {
+            DPrintf(0, "[%s] ExtendedNBStorage allocation failed \n", __FUNCTION__);
+            return false;
+        }
+        if (!BorrowPages(uPages))
```

Comment:
Can be kind of simpler:
RtlZeroMemory(m_ExtraNBStorage, ..);
m_ExtraNBStorage->m_UsedPagesCount = uPages;
m_ParentNBL->GetParentTXPath()->BorrowPages(&m_ExtraNBStorage)
Then TXPath need only 2 public members:
bool BorrowPages(CExtendedNBStorage *);
void ReturnPages(CExtendedNBStorage *);

Another option:
TXPath has only one public: CRawPageList& ExtraPages(); 

Then we can make more procedures propected/private without redundant visibility.

Reply:
@ybendito ,thanks, I'll make some changes. Actually, I initially used Borrow and Return in Tx as well. Later on, I considered that doing so would cause some coupling between CExtendedNBStorage（I'm debating whether CExtendedNBStorage should only be tightly coupled with CNB.） and CParaNdisTX, and after some hesitation, I changed it to the current form))

Reply:
@ybendito ,Changed to bool BorrowPages(CExtendedNBStorage *);
void ReturnPages(CExtendedNBStorage *);
I've realized an issue, that even in this way, within the NB class, we still can't avoid exposing ReturnPages(), because TXPath won't record the pages that have already been used. That's my understanding, is that correct?